### PR TITLE
s3: Allow turning off checksums in S3 Puts

### PR DESCRIPTION
**Pull request by sinkingpoint**
https://github.com/thanos-io/thanos/pull/6746 bumped objstore, which bumped minio, which made Thanos incompatible with otherwise compliant S3 backends that do not support the x-amz-checksum header.

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

- Add in a `PutWithMD5` (open to naming changes) config entry to the S3 bucket config that allows turning off this header so that Thanos supports these backends.

## Verification

Running this at Cloudflare on our [r2](https://developers.cloudflare.com/r2/) backed Thanos that doesn't support this header

### CODE REVIEW

**Comment by sinkingpoint** [comment]
@yeya24 @andyasp Mind giving this a look when you get a sec?

**Review by yeya24 — APPROVED** [review]
Generally it looks good. I am okay with the naming but let's see if other maintainers have other suggestions.

It would be better if we can have some doc/comment saying what is this configuration for and what S3 compatible backends might need to turn on/off this flag.

**Comment by douglascamata** [comment]
I am slightly worried that there could be many people out there facing this problem and they won't easily notice. 

I think we should include the md5 always and by default to try to avoid these situations. I personally don't foresee any issues with this.

Do you think it could cause problems, @sinkingpoint @yeya24?

**Comment by mlech-reef** [comment]
This problem also exists for Backblaze B2. This change would be super-useful as currently we can not use Thanos with B2 and multipart.

Forcing MD5 is also ok for us, but I guess someone that use AWS S3 could benefit from `x-amz-checksum-*` feature.

**Comment by mlech-reef** [comment]
IMHO `SendContentMD5` as it is called in minio is more consistent with AWS S3 documentation - https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html#checking-object-integrity-md5

**Comment by sinkingpoint** [comment]
Updated the name to be the same as minio. That makes sense - its a lot more explicit

**Comment by sinkingpoint** [comment]
Defaulting to true, to disable the header by default seems like a good idea - technically speaking that would restore the previous behavior because otherwise the prior PR was a breaking change

**Comment by yeya24** [comment]
I see it is still set to false. Do we want to make it default to true before I merge this

**Comment by douglascamata** [comment]
I'm not a big fan of how they handle these things over at Minio: https://github.com/minio/minio-go/issues/1791. I'm 100% in favor of this project keeping the compatibility with as many s3-compatible object storage providers as possible (that's the whole purpose) without requiring users to manually disable some "random" flag because it messes up everything else besides s3 and minio itself..

**Comment by sinkingpoint** [comment]
Added the `true` default, and a CHANGELOG entry

**Review by douglascamata — APPROVED** [review]

**Review by yeya24 — APPROVED** [review]

**Comment by adamlounds** [comment]
for reference, Cloudflare r2 doesn't seem to need this any more

**Pull request merged**
